### PR TITLE
Fix some theoretical problems with png reading

### PR DESCRIPTION
**Pull request by mdboom**
Possible fix fo #5495, but honestly quite theoretical.

### CODE REVIEW

**Comment by WeatherGod** [comment]
So, if for some reason the read method isn't there (I can't think of a reason why), how does the caller know that something went wrong?

**Comment by mdboom** [comment]
There's no way to report an error back from a PNG reading callback.  However, when the PNG reading is finished, we check `PyErr_Occurred()` and send the exception back to the caller.

https://github.com/matplotlib/matplotlib/blob/master/src/_png.cpp#L576

In this case, the user would get an AttributeError with whatever message `PyObject_CallFunction` set.

File: src/_png.cpp [hunk 1]
```diff
@@ -328,10 +328,10 @@ static void _read_png_data(PyObject *py_file_obj, png_bytep data, png_size_t len
...
     if (read_method) {
         result = PyObject_CallFunction(read_method, (char *)"i", length);
-    }
-    if (PyBytes_AsStringAndSize(result, &buffer, &bufflen) == 0) {
-        if (bufflen == (Py_ssize_t)length) {
-            memcpy(data, buffer, length);
+        if (PyBytes_AsStringAndSize(result, &buffer, &bufflen) == 0) {
+            if (bufflen == (Py_ssize_t)length) {
+                memcpy(data, buffer, length);
+            }
```
Comment (WeatherGod) [inline src/_png.cpp]: Ok, I see that now. Now, what if `bufflen` does not end up being the same as `length`? Are we assuming that an error has been set? Would it make sense for us to set an error if one hasn't been set?

Comment (mdboom) [inline src/_png.cpp]: Good point.  We should probably set an exception in that case.  It basically means that it read past the end of the file (which could happen if the file were truncated or something).

**Comment by tacaswell** [comment]
back ported to v2.0.x as 3814c77823806d0047d08fa3cbfb822f24b33130

**Comment by QuLogic** [comment]
Not 1.5.x (as milestoned)?

**Pull request merged**